content/post/policy-based-design-cpp.md

```diff
@@ -0,0 +1,179 @@
++++
+
+title = "Policy based design in C++"
+author = "vmrob"
+draft = false
+date = "2015-09-01T19:05:13-06:00"
+tags = ["C++"]
+description = "Impementing the strategy pattern through the use of templatized mixins in C++"
+
++++
+
+### Design Patterns
+
+As software developers, we use design patterns to name and classify common programming idioms using a common set of jargon. It allows us to describe the higher level design of our code while providing a good idea of what a set of code accomplishes without having to understand everything about it.
+
+One such design pattern, the [strategy pattern](https://en.wikipedia.org/wiki/Strategy_pattern), is quite familiar to those of us that use the C++ Standard Template Library (STL). The pattern is essentially the encapsulation of an algorithm so that the programmer can specify the behavior of a separate class or algorithm in a dynamic way. A great example used by all of the ordered STL containers is the Compare template argument--it defines how the container determines the total ordering of its elements.
+
+```C++
+template < typename T,                   // key_type/value_type
+           typename Compare = less<T>,   // key_compare/value_compare
+           typename Alloc = allocator<T> // allocator_type
+           > class set;
+```
+
+Other simple uses of the strategy pattern are the use of predicate functions provided to the C++ `<alogrithm>` functions.
+
+```C++
+std::string str = "Foo Bar";
+
+str.erase(std::remove_if(str.begin(), str.end(),
+    [](char c) {
+        return std::isspace(c) != 0;
+    }), str.end());
+
+std::cout << str << std::endl; // "FooBar"
+```
+
+In the above example, the logic of _what_ to erase has been separated from the logic of _how_ it is erased. The result is a completely reusable set of functions that can be mixed and matched to cover even more functionality.
+
+### Mixins
+
+Many OOP languages have the concept of [mixins](https://en.wikipedia.org/wiki/Mixin) which promote the simplification of object design by reducing a class to a series of reusable data and behavior. In php, these are called traits and can be used to keep classes small and tightly focused.
```

Comment:
Yeah. I should. Same with other things we normally use in a lowercase context like Git.